substrate/client/network/src/litep2p/shim/notification/tests/fuzz.rs

```diff
@@ -0,0 +1,373 @@
+
+use std::{
+	collections::{HashMap, HashSet},
+	sync::Arc,
+};
+
+#[tokio::test]
+#[cfg(debug_assertions)]
+async fn run() {
+	sp_tracing::try_init_simple();
+
+	for _ in 0..50 {
+		test_once().await;
+	}
+}
+
+#[cfg(debug_assertions)]
+async fn test_once() {
+	// PRNG to use.
+	let mut rng = rand::thread_rng();
+
+	// peers that the peerset knows about.
+	let mut known_peers = HashSet::<PeerId>::new();
+
+	// peers that we have reserved. Always a subset of `known_peers`.
+	let mut reserved_peers = HashSet::<PeerId>::new();
+
+	// reserved only mode
+	let mut reserved_only = Uniform::new_inclusive(0, 10).sample(&mut rng) == 0;
+
+	// Bootnodes for `PeerStore` initialization.
+	let bootnodes = (0..Uniform::new_inclusive(0, 4).sample(&mut rng))
+		.map(|_| {
+			let id = PeerId::random();
+			known_peers.insert(id);
+			id
+		})
+		.collect();
+
+	let peerstore = Peerstore::new(bootnodes);
+	let peer_store_handle = peerstore.handle();
+
+	let (mut peerset, to_peerset) = Peerset::new(
+		ProtocolName::from("/notif/1"),
+		Uniform::new_inclusive(0, 25).sample(&mut rng),
+		Uniform::new_inclusive(0, 25).sample(&mut rng),
+		reserved_only,
+		(0..Uniform::new_inclusive(0, 2).sample(&mut rng))
+			.map(|_| {
+				let id = PeerId::random();
+				known_peers.insert(id);
+				reserved_peers.insert(id);
+				id
+			})
+			.collect(),
+		Default::default(),
+		Arc::clone(&peer_store_handle),
+	);
+
+	tokio::spawn(peerstore.run());
+
+	// opening substreams
+	let mut opening = HashMap::<PeerId, Direction>::new();
+
+	// open substreams
+	let mut open = HashMap::<PeerId, Direction>::new();
+
+	// closing substreams
+	let mut closing = HashSet::<PeerId>::new();
+
+	// closed substreams
+	let mut closed = HashSet::<PeerId>::new();
+
+	// perform a certain number of actions while checking that the state is consistent.
+	//
+	// if we reach the end of the loop, the run has succeeded
+	let _ = tokio::task::spawn_blocking(move || {
+		// PRNG to use in `spawn_blocking` context.
+		let mut rng = rand::thread_rng();
+
+		for _ in 0..2500 {
+			// each of these weights corresponds to an action that we may perform
+			let action_weights =
+				[300, 110, 110, 110, 110, 90, 70, 30, 110, 110, 110, 110, 20, 110, 50];
+
+			match WeightedIndex::new(&action_weights).unwrap().sample(&mut rng) {
+				0 => match peerset.next().now_or_never() {
+					// open substreams to `peers`
+					Some(Some(PeersetNotificationCommand::OpenSubstream { peers })) =>
+						for peer in peers {
+							opening.insert(peer, Direction::Outbound);
+							assert!(!closing.contains(&peer));
+							assert!(!open.contains_key(&peer));
+						},
+					// close substreams to `peers`
+					Some(Some(PeersetNotificationCommand::CloseSubstream { peers })) =>
+						for peer in peers {
+							assert!(closing.insert(peer));
+							if open.remove(&peer).is_none() {
+								panic!("peer {peer:?} doesn't exist in `open`");
+							}
+							// assert!(open.remove(&peer).is_some());
```

Comment:
```suggestion
```

Reply:
Decide to remove the `panic!()` instead since it's slightly cleaner and the panic was just used to debug an issue in `Peerset`